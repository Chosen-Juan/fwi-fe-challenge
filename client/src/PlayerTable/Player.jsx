import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Flags from 'react-world-flags';
import { connect } from 'react-redux'

import Avatar from '../Avatar';
import { COUNTRIES, COUNTRIES_ARRAY } from '../constants';
import { editPlayer, deletePlayer } from '../appState/actions';

export class Player extends Component {
  static propTypes = {
    player: PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      country: PropTypes.oneOf(Object.keys(COUNTRIES)),
      winnings: PropTypes.number.isRequired,
      imageUrl: PropTypes.string.isRequired,
    }),
    editPlayer: PropTypes.func.isRequired,
    deletePlayer: PropTypes.func.isRequired,
  };

  handleButtonPress = () => {
    this.buttonPressTimer = setTimeout(() => this.setEditing(true), 1000);
  };

  handleButtonRelease = () => {
    clearTimeout(this.buttonPressTimer);
  };

  state = {
    editing: false,
    name: this.props.player.name,
    winnings: this.props.player.winnings,
    country: this.props.player.country,
  };

  setEditing = editing => {
    this.setState({ editing });
  };

  deletePlayer = () => {
    const { deletePlayer, player } = this.props;
    deletePlayer(player.id);
  };

  saveEdit = () => {
    const { id } = this.props.player;
    const { name, winnings, country } = this.state;
    this.props.editPlayer(id, name, parseInt(winnings), country)
      .then(this.finishEditing);
  };

  finishEditing = () => {
    this.setState({ success: true });
    setTimeout(() => {
      this.setState({ success: false });
      this.setEditing(false);
    }, 3000);
  };

  onChangeHandler = ({ target }) => {
    this.setState({ [target.name]: target.value });
  };

  renderCountryOptions = () => {
    const options = [
      <option key="default" disabled value>Select a country</option>
    ];
    COUNTRIES_ARRAY.forEach(country => {
      options.push(
        <option key={country.code} value={country.code}>
          {country.value}
        </option>
      );
    });
    return options;
  }

  render() {
    const { player } = this.props;
    return (
      <tr
        key={player.id}
        role="row"
        className="table__row"
        onTouchStart={this.handleButtonPress}
        onTouchEnd={this.handleButtonRelease}
      >
        <td className={this.state.editing ? "is-hidden-desktop" : "is-hidden"}>
          <button className="button is-primary" onClick={this.saveEdit}>Save</button>
          <button className="button is-danger" onClick={() => this.setEditing(false)}>Cancel</button>
        </td>
        <td className="is-narrow">
          <Avatar src={player.imageUrl} />
        </td>
        <td hidden={this.state.editing}>
          {player.name}
        </td>
        <td hidden={!this.state.editing}>
          <input name="name" type="text" className="input" value={this.state.name} onChange={this.onChangeHandler} />
        </td>
        <td hidden={this.state.editing}>
          {player.winnings.toLocaleString(undefined, {
            style: 'currency',
            currency: 'USD',
          })}
        </td>
        <td hidden={!this.state.editing}>
          <input name="winnings" type="number" className="input" value={this.state.winnings} onChange={this.onChangeHandler} />
        </td>
        <td hidden={this.state.editing}>
          <div className="country">
            <Avatar>
              <Flags code={player.country} alt="" />
            </Avatar>
            {player.country}
          </div>
        </td>
        <td hidden={!this.state.editing}>
          <select name="country" className="input" value={this.state.country} onChange={this.onChangeHandler}>
            {this.renderCountryOptions()}
          </select>
        </td>
        <td className={this.state.success ? "notification is-success" : "is-hidden"}>Saved</td>
        <td className={this.state.success ? "is-hidden" : "is-narrow"}>
          <div className={this.state.editing ? "is-hidden-touch": "is-hidden"}>
            <button className="button is-primary" onClick={this.saveEdit}>Save</button>
            <button className="button is-danger" onClick={() => this.setEditing(false)}>Cancel</button>
          </div>
          <button className={this.state.editing ? "is-hidden" : "button is-link is-outlined is-hidden-touch"} onClick={() => this.setEditing(true)}>Edit</button>
          <button className={this.state.editing ? "is-hidden" : "button is-danger is-outlined"} onClick={this.deletePlayer}>&times;</button>
        </td>
      </tr>
    );
  }
}
const mapStateToProps = state => ({
  playerStore: state.player
});

const mapDispatchToProps = {
  editPlayer, deletePlayer
};

export default connect(mapStateToProps, mapDispatchToProps)(Player);
