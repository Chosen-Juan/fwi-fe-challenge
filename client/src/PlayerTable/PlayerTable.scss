@import 'variables';

$header-height: 2.5rem;

.table {
  margin-top: 3.5rem;
  padding-top: calc($header-height + 1px); // 1px for border of header
  width: 100%;
  margin-bottom: 0 !important;


  &__row {
    background-color: $snow;
    transition: background-color 0.15s;

    &:hover {
      background-color: $pearl;
    }

    &:not(:last-child) {
      border-bottom: 1px solid $pearl;
    }
  }

  .centered-text {
    vertical-align: inherit;
  }
}

.country {
  align-items: center;
  display: flex;
  height: 100%;

  .avatar {
    margin-right: 0.5rem;
  }
}
