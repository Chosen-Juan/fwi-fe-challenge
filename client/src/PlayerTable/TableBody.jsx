import React from 'react';
import PropTypes from 'prop-types';
import Flags from 'react-world-flags';

import Avatar from '../Avatar';
import { COUNTRIES } from '../constants';

const TableBody = ({ players }) => {
  return (
    <tbody className="table--body">
      {players.map(({ id, name, country, winnings, imageUrl }) => (
        <tr key={id} role="row" className="table__row">
          <td>
            <Avatar src={imageUrl} />
          </td>
          <td className="centered-text">
            {name}
          </td>
          <td className="centered-text">
            {winnings.toLocaleString(undefined, {
              style: 'currency',
              currency: 'USD',
            })}
          </td>
          <td>
            <div className="country">
              <Avatar>
                <Flags code={country} alt="" />
              </Avatar>
              {country}
            </div>
          </td>
        </tr>
      ))}
    </tbody>
  );
};

TableBody.propTypes = {
  players: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      country: PropTypes.oneOf(Object.keys(COUNTRIES)),
      winnings: PropTypes.number.isRequired,
      imageUrl: PropTypes.string.isRequired,
    })
  ).isRequired,
};

export default TableBody;
