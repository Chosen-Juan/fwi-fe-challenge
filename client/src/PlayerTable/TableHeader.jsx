import React from 'react';

const TableHeader = () => (
  <thead>
    <tr>
      <th />
      <th>
        Player
      </th>
      <th>
        Winnings
      </th>
      <th>
        Native of
      </th>
    </tr>
  </thead>
);

export default TableHeader;
