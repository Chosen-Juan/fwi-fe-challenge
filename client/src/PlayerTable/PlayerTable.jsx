import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connectAdvanced } from 'react-redux';
import shallowEqual from 'shallowequal';

import { COUNTRIES } from '../constants';
import { fetchPlayers } from '../appState/actions';

import './PlayerTable.scss';
import TableHeader from './TableHeader';
import TableBody from './TableBody';
import Pages from '../Pages/Pages';

class PlayerTable extends PureComponent {
  static propTypes = {
    players: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        country: PropTypes.oneOf(Object.keys(COUNTRIES)),
        winnings: PropTypes.number.isRequired,
        imageUrl: PropTypes.string.isRequired,
      })
    ).isRequired,
    fetchPlayers: PropTypes.func.isRequired,
  };

  componentDidMount() {
    this.props.fetchPlayers();
  }

  render() {
    const { players } = this.props;
    return (
      <div className="player-table">
        <Pages />
        <table
          aria-label="Poker Players"
          className="table is-scrollable"
        >
          <TableHeader />
          <TableBody players={players} />
        </table>
      </div>
    );
  }
}

export default connectAdvanced(dispatch => {
  let result;
  const actions = bindActionCreators({ fetchPlayers }, dispatch);

  return (state, props) => {
    const players = state.playerIds.map(id => state.players[id]);
    const { from, size, total } = state;
    const nextResult = { ...props, ...actions, players, from, size, total };

    if (!shallowEqual(result, nextResult)) {
      result = nextResult;
    }

    return result;
  };
})(PlayerTable);
