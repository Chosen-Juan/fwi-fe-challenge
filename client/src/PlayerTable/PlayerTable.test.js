import React from 'react';
import { shallow } from 'enzyme';

import { PlayerTable } from './PlayerTable';

describe('PlayerTable', () => {
  const fetchPlayersMock = jest.fn();
  const defaultProps = {
    players: [{
      id: '1one',
      name: 'Jon Doe',
      country: 'US',
      winnings: 24999,
      imageUrl: ''
    }],
    fetchPlayers: fetchPlayersMock
  };
  const playerTable = shallow(<PlayerTable {...defaultProps} />);

  it('should match snapshot', () => {
    expect(playerTable).toMatchSnapshot();
  });

  it('should have called fetch players', () => {
    expect(fetchPlayersMock).toHaveBeenCalledTimes(1);
  });

});
