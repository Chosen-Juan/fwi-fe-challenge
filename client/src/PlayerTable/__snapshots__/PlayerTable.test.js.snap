// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerTable should match snapshot 1`] = `
<div
  className="player-table"
>
  <Connect(Pages) />
  <table
    aria-label="Poker Players"
    className="table is-scrollable"
  >
    <TableHeader />
    <TableBody
      players={
        Array [
          Object {
            "country": "US",
            "id": "1one",
            "imageUrl": "",
            "name": "Jon Doe",
            "winnings": 24999,
          },
        ]
      }
    />
  </table>
</div>
`;
